docs/authentication_architecture.md

```diff
@@ -0,0 +1,71 @@
+can tell who opened the connection and that it was intended for this instance of **JF**.
+
+Problems:
+
+- who creates the token? (more on that later),
+- the token can expire (this doesn's seem like a problem, when the token is used only to open WebSocker connection, but user possibly needs to implement logic to refresh the token).
+
+### Using **BE** as a middleman
+
+**Note: direct signaling approach has been chosen, so this paragraph is not applicable anymore.**
+
+The same assumptions and first 3 steps as in the previous example:
+
+4) **BE** responds to **CL** that it was authorized and that it can send signaling messages to **BE**.
+
+Problems:
+
+- in current implementation of `membrane_rtc_engine`, endpoint should be created after signaling connection was established. In previous example it was obvious:
+WebSocekt connection is established, endpoint is created. Here some kind of special message will be necessary to communicate the begining of signaling connection. Also, we need
+to think if we want to use the token to tag the signaling messages (seems unnecessary), or simply `peer_id`, it's harder to manage permissions
+(more of a signaling architecture problems, but worth noting),
+- **BE** doesn't need to pass the token to **CL** (has to respond to in anyway to start signaling messages flow), but have to take care of matching incoming messages with `peer_id`s.~
+
+## Who should create tokens?
+
+### **BE**
+
+If **BE** (server SDK) is responsible for creating tokens, then it has to know the secret that will be used in signing JWT. It also needs `peer_id`, which can be obtained from **JF**. That takes care of **CL** authorization, but we also need to authorize **BE** - **JF** connection. One possible solution is to use the very same secret as before to create tokens that will be used by **BE**. So **BE** will generate token and then, instead of sending them to client, will use it itself (which may seem wierd, but we came to conclusion that the approach is alright).
+
+This approach only makes sense when using direct signaling, otherwise tokens are not necessary at all (except for **BE** - **JF** connection).
+
+### **JF**
+
+In this approach **JF** creates the tokens (when using `add_peer`, token is send in the response), so user only needs to pass it to the **CL**.
+You also might need to pass expected token presmissions to **JF**.
+Despite that, we still need a way to authenticate **BE** - **JF** connection. Possible solutions:
+
+- create JWT on **BE** side anyway (in such case you might as well use the first approach in order not to split logic responsible for token generation between Server SDK and Jellyfish),
+- create JWT (or some other token type) once and use it in configuration (makes it easier to change **BE** permissions, if that's ever necessary, but the token never expires, I'm not sure whether that's a problem, also **BE** doesn't need to know the secret).
```

Comment:
We probably don't want to implement that in the 0.1.0, but we should think about secret rotation. Some security certifications will require the secrets to be rotated once a few months or in case of a security breach.

Also, I think I would prefer an option where you can limit the permissions of the BE. For once, applications should always have a set of permissions that gives them access only to the features that they actually use. But even more importantly, if we implemented secret rotation, the permissions for this should only be given to the process that performs the actual rotation.  And if the BE has a token with limited permissions, it cannot know the secret used to sign that token - if it did, it could generate a new token with higher permissions and then we might as well not bother about limiting the permissions.

Reply:
> And if the BE has a token with limited permissions, it cannot know the secret used to sign that token - if it did, it could generate a new token with higher permissions and then we might as well not bother with limiting the permissions.

Any ideas how to achieve this?

To me it sounds like we don't trust our own backend and would like to have some other security entity over it